### PR TITLE
Update offline distillation and saving top-k teacher logits to be efficient

### PR DESCRIPTION
**Pull request by ajkv-google**
# Description

This PR implements an optimized Offline Distillation pipeline in MaxText, allowing student models to train using pre-saved teacher logits to significantly reduce compute costs. 

Key changes:

- The `save_top_k_teacher_logits.py` script now supports parallel execution across hosts and uses asynchronous GCS uploads to prevent TPU idling 
- Offline distillation now uses grain input pipeline without having to use the custom offline arrayrecord iterator. This speeds up the offline distillation process without having to read logits on a single thread. 
- Introduced sparse KL divergence in `distillation_utils.py` that computes KL divergence using only the teacher's top-k predictions when running offline distillation

# Tests

Used the following to run offline distillation:

- Command: https://paste.googleplex.com/4852897526448128
- Yaml: https://paste.googleplex.com/4981440528908288
- Example run for offline distillation (using our team's LTI checkpoint for 30b model): [Tensorboard Link](https://pantheon.corp.google.com/storage/browser/ajkv-distillation/offline-distillation-runs/qwen3-30b/experiments-1500/lti-our-30b/qwen3-30b-a3b-base_2026-05-22-05-19/tensorboard?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&e=13802955&mods=allow_workbench_image_override)

Used the following to run saving of top-k teacher logits:

- Command: https://paste.googleplex.com/6383382496935936
- Yaml: https://paste.googleplex.com/5898105013796864
- GS Bucket Path for saved logits (1B tokens of climbmix): [Saved offline logits](https://pantheon.corp.google.com/storage/browser/ajkv-distillation/teacher-logits-run/qwen3-30b/1500_logits/d-ajkv-16533;tab=objects?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev&prefix=&forceOnObjectsSortingFiltering=false)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 1]
```diff
@@ -538,32 +495,53 @@ def compute_loss(
...
-    # KL(teacher || student) per position. optax.kl_divergence(log_pred, target) = KL(target || pred).
-    kl_softened_per_pos = optax.kl_divergence(log_s_T, t_p_T)  # [B, T]
+    if getattr(teacher_output, "top_k_indices", None) is not None:
+      # --- SPARSE KL DIVERGENCE (Offline Mode) ---
+      # 1. Normalize teacher probabilities ONLY over the saved Top-K subset
+      t_p_T_sparse = jax.nn.softmax(t_logits / temperature, axis=-1)
+      log_t_p_T_sparse = jax.nn.log_softmax(t_logits / temperature, axis=-1)
+
+      # 2. Student log-probs must be computed over the FULL vocabulary to be mathematically valid
+      log_s_T_full = jax.nn.log_softmax(s_logits / temperature, axis=-1)
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py]: should we also compute teacher logits softmax over entire vocabulary before saving to files?

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 2]
```diff
@@ -538,32 +495,53 @@ def compute_loss(
...
     valid_count = jnp.sum(mask)
     safe_count = jnp.maximum(valid_count, 1.0)
-
+    
     # --- Soft loss: KL on temperature-softened distributions ---
-    log_s_T = jax.nn.log_softmax(s_logits / temperature, axis=-1)
-    t_p_T = jax.nn.softmax(t_logits / temperature, axis=-1)
-    # KL(teacher || student) per position. optax.kl_divergence(log_pred, target) = KL(target || pred).
-    kl_softened_per_pos = optax.kl_divergence(log_s_T, t_p_T)  # [B, T]
+    if getattr(teacher_output, "top_k_indices", None) is not None:
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py:503]: there are many repeating code lines for sparse vs regular loss compute, looks like it can be refactored to reduce amount of changed code

Comment (ajkv-google) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py:503]: Yes, there are repeated lines of code, so I removed the redundancy and made the code cleaner.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 1 above)
Comment (ajkv-google) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: We save the raw logits instead of the full softmax so we can still tweak the temperature on the fly during training. Also, saving full-vocab probabilities would make the files bigger, which could reduce read speeds during offline training. I think as of now, it would be good to stick to the current approach where we store the information that is needed, and read it quickly during the offline training.

Comment (vlad-karp) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py]: "lso, saving full-vocab probabilities would make the files bigger" - why do you consider saving the full vocab? the entire idea of offline is to use a limited set of logits.
My concern is purely mathematical - you normalize only over top-k while the student logits will be normalized over entire vocabulary, and then you calculate kl divergence over those distributions with completely different normalization scales.

File: src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py [hunk 3]
```diff
@@ -91,154 +160,142 @@ def generate_and_save_data(config, local_args):
...
   steps_per_file = local_args.steps_per_file
 
+  writer = None
+  local_output_path = None
+
+  # all hosts initialize their own directories and thread pools
+  if not os.path.exists(local_tmp_dir):
+    os.makedirs(local_tmp_dir, exist_ok=True)
+
+  upload_executor = ThreadPoolExecutor(max_workers=4)
```
Comment (JamesDeng42) [inline src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py:170]: the concern i have is that if we do not cap total number of threads (now this is 4*total_number_of_chips i think), it might choke network bandwidth. we can probably set it to be single thread by default to be on the conservative side, if we need performance boost, then set it to be multi-thread (2 or 4).

**Review by JamesDeng42 — APPROVED** [review]
In general, LGTM

**Review by vlad-karp — APPROVED** [review]

File: src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py (hunk 3 above)
Comment (ajkv-google) [reply src/maxtext/trainers/post_train/distillation/save_top_k_teacher_logits.py:170]: That's a good point. I could make this a command line arg in the future PR to improve this.
